agent/containermetadata/parse_metadata.go

```diff
@@ -0,0 +1,137 @@
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package containermetadata
+
+import (
+	"fmt"
+
+	"github.com/aws/amazon-ecs-agent/agent/api"
+
+	"github.com/cihub/seelog"
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+// parseMetadataAtContainerCreate gathers metadata from task and cluster configurations
+// then packages it for JSON Marshaling. We use this version to get data
+// available prior to container creation
+// Since we accept incomplete metadata fields, we should not return
+// errors here and handle them at this or the above stage.
+func (manager *metadataManager) parseMetadataAtContainerCreate(taskARN string, containerName string) Metadata {
+	return Metadata{
+		cluster:              manager.cluster,
+		taskMetadata:         TaskMetadata{containerName: containerName, taskARN: taskARN},
+		containerInstanceARN: manager.containerInstanceARN,
+		metadataStatus:       MetadataInitial,
+	}
+}
+
+// parseMetadata gathers metadata from a docker container, and task
+// configuration and data then packages it for JSON Marshaling
+// Since we accept incomplete metadata fields, we should not return
+// errors here and handle them at this or the above stage.
+func (manager *metadataManager) parseMetadata(dockerContainer *docker.Container, taskARN string, containerName string) Metadata {
+	dockerMD := parseDockerContainerMetadata(taskARN, containerName, dockerContainer)
+	return Metadata{
+		cluster:                 manager.cluster,
+		taskMetadata:            TaskMetadata{containerName: containerName, taskARN: taskARN},
+		dockerContainerMetadata: dockerMD,
+		containerInstanceARN:    manager.containerInstanceARN,
+		metadataStatus:          MetadataReady,
+	}
+}
+
+// parseDockerContainerMetadata parses the metadata in a docker container
+// and packages this data for JSON marshaling
+// Since we accept incomplete metadata fields, we should not return
+// errors here and handle them at this stage.
+func parseDockerContainerMetadata(taskARN string, containerName string, dockerContainer *docker.Container) DockerContainerMetadata {
+	if dockerContainer == nil {
+		seelog.Warnf("Failed to parse container metadata for task %s container %s: container metadata not available or does not exist", taskARN, containerName)
+		return DockerContainerMetadata{}
+	}
+
+	// In most cases a container should never lack a config but we check regardless to avoid
+	// nil pointer exceptions (Could occur if there is some error in the docker api call, if the
+	// container we receive has incomplete information)
+	imageNameFromConfig := ""
+	if dockerContainer.Config != nil {
+		imageNameFromConfig = dockerContainer.Config.Image
+	} else {
+		seelog.Warnf("Failed to parse container metadata for task %s container %s: container has no configuration", taskARN, containerName)
+	}
+
+	// Get Port bindings from docker configurations
+	var ports []api.PortBinding
+	var err error
```

Comment:
minor: if there's no specific reason to declare this here, can you get rid of it?

Reply:
im going to leave this one as it is, the nested branch needs both, and we return ports even if its empty.